xds/src/main/java/io/grpc/xds/sds/SdsSslContextProvider.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import io.envoyproxy.envoy.api.v2.auth.CertificateValidationContext;
+import io.envoyproxy.envoy.api.v2.auth.CommonTlsContext;
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.auth.SdsSecretConfig;
+import io.envoyproxy.envoy.api.v2.auth.Secret;
+import io.envoyproxy.envoy.api.v2.auth.TlsCertificate;
+import io.envoyproxy.envoy.api.v2.auth.UpstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.core.Node;
+import io.grpc.Status;
+import io.grpc.netty.GrpcSslContexts;
+import io.grpc.xds.sds.trust.SdsTrustManagerFactory;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import java.io.IOException;
+import java.security.cert.CertStoreException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * An SslContext provider that uses SDS to fetch secrets. Used for both server and client
+ * SslContexts
+ */
+final class SdsSslContextProvider<K> extends SslContextProvider<K>
```

Comment:
there isn't much code sharing for client and server. can we split? it seems like combining is just adding more complexity for not much gain.

Reply:
Most code is common. Only `updateSslContext()` has 2 distinct `if-else` blocks for client vs server (that too because `GrpcSslContexts` forces me to call `forServer` or `forClient` instead of passing a flag) so not worth splitting into 2 classes for that?

Reply:
resolving for now since I don't think splitting is a good idea.